### ISSUE TITLE
Automatic refresh of access tokens not working when using omniauth-google-oauth2 gem

### ISSUE THREAD

**Issue by karianne** [claudiob/yt#189]
I'm struggling to get automatic refresh of access tokens to work. I'm on the newest version of the Yt gem, and I'm using the `omniauth-google-oauth2` gem and Rails 4.2 as well. This is the code and what it outputs:

```
2.2.2 (main):0 > account = Yt::Account.new refresh_token: <my_token>
=> #<Yt::Models::Account:0x007fc734007e88
@access_token=nil,
@authentication=nil,
@authorization_code=nil,
@device_code=nil,
@expires_at=nil,
@force=nil,
@redirect_uri=nil,
@refresh_token="<my_token>",
@scopes=nil>
2.2.2 (main):0 > account.email
Yt::Errors::Unauthorized: A request to YouTube API was sent without a valid authentication:
{}

You can retry the same request manually by running:



from /Users/karianneberg/code/yt/lib/yt/associations/has_authentication.rb:91:in `rescue in use_refresh_token!'
```

With these initializers:

```
Yt.configure do |config|
   config.api_key = ENV['GOOGLE_API_KEY']
   config.log_level = :devel
 end

Rails.application.config.middleware.use OmniAuth::Builder do
  provider :google_oauth2,
       ENV['GOOGLE_CLIENT_ID'],
       ENV['GOOGLE_CLIENT_SECRET'],
       scope: 'userinfo.profile,youtube.readonly,yt-analytics.readonly,youtubepartner-content-owner-readonly,youtubepartner',
       prompt: 'consent'
end
```

Please note that the client id and secret are set on the Omniauth middleware, not in `Yt.configure`. This seems to work for everything I've tried to do so far (fetching videos, fetching channels, fetching stats etc) except the refreshing of access tokens using refresh token. A workaround is to set the id and secret directly in `Yt.configure` as well, but that seems redundant. 

Is this intended behaviour, or a bug?

**Comment by stavro** [comment]
Can you confirm that you have your `client_id` and `client_secret` configured inside of Yt?

```
Yt.configure do |config|
  config.client_id = "..."
  config.client_secret = "..."
end

```

**Comment by karianne** [comment]
@stavro Id and secret are set. If I set an access token on the account, it works until it expires.

**Comment by karianne** [comment]
I did some more digging and debugging, and it seems like `List#first!` raises due to `List#list` returning an enumerator with no items. I have no idea where those items are supposed to come from - a call to the `/o/oauth2/token` endpoint has not been made.

**Comment by karianne** [comment]
I figured out what was happening. Issue is updated.

**Comment by stavro** [comment]
Hi @karianne!  Thanks for reaching out.

> Please note that the client id and secret are set on the Omniauth middleware, not in Yt.configure. This seems to work for everything I've tried to do so far (fetching videos, fetching channels, fetching stats etc) except the refreshing of access tokens using refresh token.

This is because you only need a valid access token to interact with the majority of the YouTube API.  Once a user OAuth's with you for the first time, you get a token which lasts an hour.  Yt may be used within this hour time period until the token expires, which you will then need to refresh.... Which leads to your problem.

> A workaround is to set the id and secret directly in Yt.configure as well, but that seems redundant.

I can see how you feel it is redundant, but you must keep in mind that **Yt** and **OmniAuth** are completely isolated and distinct libraries.  As the creators of Yt, we must not assume anything about your application environment, including whether or not a user of Yt also is using OmniAuth.  As such, Yt must be configured appropriately regardless of whatever other gems or packages you are using.

It is not a _workaround_ to configure Yt in addition to OmniAuth, but rather it is mandatory basic gem configuration.

Also, unless you're using some other OAuth services (Facebook, etc), you might find it easier to just use Yt for all your OAuth needs.  Yt can do nearly all of what OmniAuth provides for your YouTube integrations, and you'll minimize your external dependencies.

**tl;dr** - Yes, you need to configure Yt with your client id and secret credentials _as well as_ whatever other gems which use YouTube APIs, such as OmniAuth.

**Comment by karianne** [comment]
Hi @stavro , thanks for clearing this up! I see that I've completely misunderstood how this works. I was not aware that the client id and secret was used only for refreshing access tokens, and I thought Yt was integrated with omniauth in some way, since most of the API was working.

It would probably have spared me quite a lot of time if the error message I received would have said something about that client id and secret needed to be configured, instead of just giving me a generic error message like that. As far as I can understand, a request to the Youtube API has in fact _not_ been made yet at the time that the error is raised, so this is perhaps a bit misleading.
